src/lib/sw_engine/tvgSwRenderer.cpp

```diff
@@ -92,41 +88,38 @@ struct SwShapeTask : SwTask
             uint8_t alpha = 0;
             sdata->fillColor(nullptr, nullptr, nullptr, &alpha);
             alpha = static_cast<uint8_t>(static_cast<uint32_t>(alpha) * opacity / 255);
-            bool renderShape = (alpha > 0 || sdata->fill());
-            if (renderShape || validStroke) {
+            visibleFill = (alpha > 0 || sdata->fill());
+            if (visibleFill || visibleStroke) {
                 shapeReset(&shape);
                 if (!shapePrepare(&shape, sdata, transform, clipRegion, bbox, mpool, tid)) goto err;
-                if (renderShape) {
-                    /* We assume that if stroke width is bigger than 2,
-                       shape outline below stroke could be full covered by stroke drawing.
-                       Thus it turns off antialising in that condition.
-                       Also, it shouldn't be dash style. */
-                    auto antiAlias = (strokeAlpha == 255 && strokeWidth > 2 && sdata->strokeDash(nullptr) == 0) ? false : true;
-                    if (!shapeGenRle(&shape, sdata, antiAlias, clips.count > 0 ? true : false)) goto err;
-                    filling = true;
-                }
             }
         }
 
         //Fill
         if (flags & (RenderUpdateFlag::Gradient | RenderUpdateFlag::Transform | RenderUpdateFlag::Color)) {
-            auto fill = sdata->fill();
-            if (fill) {
-                auto ctable = (flags & RenderUpdateFlag::Gradient) ? true : false;
-                if (ctable) shapeResetFill(&shape);
-                if (!shapeGenFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
-                filling = true;
-            } else {
-                shapeDelFill(&shape);
+            if (visibleFill) {
```

Comment:
The reason to handle invisible case of fill to remove the resource if I remember correctly and it looks still valid.

Reply:
please check now - it's like the old version

I'd rather do it in a similar way as it is done for a stroke - if a fill is invisible, then we don't need the shape's rle:
```
        //Fill
        if (flags & (RenderUpdateFlag::Gradient | RenderUpdateFlag::Transform | RenderUpdateFlag::Color)) {
            if (visibleFill) {
                /* We assume that if stroke width is bigger than 2,
                   shape outline below stroke could be full covered by stroke drawing.
                   Thus it turns off antialising in that condition.
                   Also, it shouldn't be dash style. */
                auto antiAlias = (strokeAlpha == 255 && strokeWidth > 2 && sdata->strokeDash(nullptr) == 0) ? false : true;
                if (!shapeGenRle(&shape, sdata, antiAlias, clips.count > 0 ? true : false)) goto err;

                if (auto fill = sdata->fill()) {
                    auto ctable = (flags & RenderUpdateFlag::Gradient) ? true : false;
                    if (ctable) shapeResetFill(&shape);
                    if (!shapeGenFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
                } else {
                    shapeDelFill(&shape);
                }
            } else {                                    // <-   the new part of the code
                rleFree(shape.rle);
                shape.rle = nullptr;
                shapeDelFill(&shape);
            }
        }
```
but since I'm not 100% sure, I left the old version...